nixos/modules/services/databases/openldap.nix

```diff
@@ -77,58 +135,101 @@ in
         example = [ "ldaps:///" ];
       };
 
+      settings = mkOption {
+        type = ldapAttrsType;
+        description = "Configuration for OpenLDAP, in OLC format";
```

Comment:
I am more a casual openldap user. Can you tell me what the advantages of the OLC format over the old format is? Most of the openldap manual that I read seems to use the old one.

Reply:
Alright. Upstream deprecates `slap.conf`.

Reply:
My other motivation was that it's more structured, for example any overlays (e.g. replication) that apply to a database are children of that database's configuration, while in the `.conf` file format everything must be specified in order. This makes mapping structured settings to OLC easier than to `.conf`. My direct motivation was that I need to run more than one database, which I can't do currently without resorting to `extraConfig`.

Another advantage is that OLC configuration can be modified without restarting the server. It is probably possible for Nix to take advantage of this, by use of a script that queries a running server for the current configuration and calculates the diff and applies that, but I have no need for this so I haven't implemented it.

Reply:
Does OLC also solve the order of acls in my ldap configuration? https://github.com/Mic92/dotfiles/blob/master/nixos/eve/modules/openldap/default.nix#L20

Reply:
This is discussed in detail [here](https://www.openldap.org/doc/admin24/slapdconf2.html), to quote:

> Some of the entries listed above have a numeric index "{X}" in their names. While most configuration settings have an inherent ordering dependency (i.e., one setting must take effect before a subsequent one may be set), LDAP databases are inherently unordered. The numeric index is used to enforce a consistent ordering in the configuration database, so that all ordering dependencies are preserved. In most cases the index does not have to be provided; it will be automatically generated based on the order in which entries are created.

For [example](https://openldap.org/doc/admin24/access-control.html):
```
olcAccess: {0}to attrs=member,entry
     by dnattr=member selfwrite
olcAccess: {1}to dn.children="dc=example,dc=com"
     by * search
olcAccess: {2}to dn.children="dc=com"
     by * read
```

If the index (`{X}`) is not specified, it is auto-generated based on the order in which the attributes are specified. This is why all members of `attrs` take a single value or a list, if a list is provided, the attributes will be inserted in order of the list, but the user can include `{X}` to override the position. I'm not sure how this interacts with file-path style values, which are included from a file at runtime though...

Reply:
Ok. Looks good to me!

Reply:
I think it's my own configuration I need to update.
In configured additional schemas here fore example:

https://github.com/Mic92/dotfiles/blob/711f8d5e20fb543267ed0620b9ee0c139675ef27/nixos/eve/modules/home-assistant/ldap.nix#L13

I need to move those first.